sql: histogram does not match column type after ALTER TABLE ALTER COLUMN SET DATA TYPE

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainGist [failed](https://tanzanite.cluster.engflow.com/invocations/default/d30c8125-4a48-445d-b587-0d955986d5c4?testReportRun=23&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on master @ [9d3dd0530818b9c5c0f7dac04e8a8de4acf1bba4](https://github.com/cockroachdb/cockroach/commits/9d3dd0530818b9c5c0f7dac04e8a8de4acf1bba4):


```
        	count(*) AS col_975
        FROM
        	defaultdb.public.seed AS tab_692
        GROUP BY
        	tab_692._jsonb
        HAVING
        	bool_and(true::BOOL)::BOOL
        ORDER BY
        	tab_692._jsonb NULLS LAST, tab_692._jsonb DESC NULLS LAST, tab_692._jsonb ASC NULLS FIRST, tab_692._jsonb NULLS LAST
        LIMIT
        	57:::INT8;
        DELETE FROM defaultdb.public.seed AS tab_693 USING defaultdb.public.seed AS tab_694 WHERE true;
        DELETE FROM defaultdb.public.seed AS tab_695 USING defaultdb.public.seed AS tab_696 WHERE false;
        INSERT
        INTO
        	defaultdb.public.seed AS tab_697
        SELECT
        	NULL AS col_976,
        	tab_698._int2 AS col_977,
        	(-2136037883700172646):::INT8 AS col_978,
        	tab_698._float8 AS col_979,
        	0.47847028619282:::FLOAT8 AS col_980,
        	'1971-04-24':::DATE AS col_981,
        	tab_698._timestamp AS col_982,
        	NULL AS col_983,
        	NULL AS col_984,
        	true AS col_985,
        	tab_698._decimal AS col_986,
        	tab_698._string AS col_987,
        	'\xdd4e':::BYTES AS col_988,
        	tab_698._uuid AS col_989,
        	tab_698._inet AS col_990,
        	tab_698._jsonb AS col_991,
        	tab_698._enum AS col_992
        FROM
        	defaultdb.public.seed AS tab_698;
        SELECT
        	NULL AS col_993, '-16 years -10 mons -69 days -21:55:19.159286':::INTERVAL AS col_994
        FROM
        	defaultdb.public.seed@seed__int8__float8__date_idx AS tab_699
        WHERE
        	true
        ORDER BY
        	tab_699._int8 ASC
        LIMIT
        	48:::INT8;
        ALTER TABLE defaultdb.public.seed ALTER COLUMN _timestamptz SET DATA TYPE TIMESTAMP;
        
    explain_test.go:205: pq: internal error: type check failed while initializing stat 977112995467198465: histogram for table seed column _timestamptz created_at 2024-06-13 06:53:41.725681 does not match column type TIMESTAMP: TIMESTAMPTZ: UPDATE defaultdb.public.seed AS tab_701 SET _int4 = 1016711112:::INT8 WHERE tab_701._bool
    --- FAIL: TestExplainGist/main (3.83s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=23</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainGist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39531

--- a/pkg/ccl/testccl/sqlccl/explain_test.go
+++ b/pkg/ccl/testccl/sqlccl/explain_test.go
@@ -200,7 +200,6 @@ func TestExplainGist(t *testing.T) {
 				// Ignore all errors except the internal ones.
 				for _, knownErr := range []string{
 					"expected equivalence dependants to be its closure",                  // #119045
-					"type check failed while initializing stat",                          // #125620
 					"argument expression has type RECORD, need type USER DEFINED RECORD", // #139910
 					"invalid datum type given: RECORD, expected RECORD",                  // #140773
 				} {

--- a/pkg/sql/logictest/testdata/logic_test/distsql_stats
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_stats
@@ -3968,3 +3968,41 @@ query T
 SELECT info FROM [EXPLAIN SELECT * FROM t155184 WHERE a < 8] WHERE info LIKE '%estimated row count:%'
 ----
   estimated row count: 4 (36% of the table; stats collected <hidden> ago)
+
+# Testcase for issue #125620: check that we can use histograms after a
+# metadata-only type conversion from TIMESTAMPTZ to TIMESTAMP.
+# The histogram should still provide accurate cardinality estimates because
+# TIMESTAMP and TIMESTAMPTZ have identical binary encodings.
+
+statement ok
+CREATE TABLE t125620 (ts TIMESTAMPTZ PRIMARY KEY)
+
+# Insert 100 rows with timestamps spanning a year.
+statement ok
+INSERT INTO t125620 SELECT '2020-01-01 00:00:00+00'::TIMESTAMPTZ + (i * INTERVAL '3 days') FROM generate_series(1, 100) AS g(i)
+
+statement ok
+ANALYZE t125620
+
+# Verify the histogram is being used before the ALTER.
+query T retry
+SELECT info FROM [EXPLAIN SELECT * FROM t125620 WHERE ts < '2020-04-01'::TIMESTAMPTZ] WHERE info LIKE '%estimated row count:%'
+----
+  estimated row count: 30 (30% of the table; stats collected <hidden> ago)
+
+statement ok
+ALTER TABLE t125620 ALTER COLUMN ts SET DATA TYPE TIMESTAMP
+
+# After the ALTER, the histogram (which stores TIMESTAMPTZ bucket bounds) should
+# still work correctly with TIMESTAMP comparisons because both types use
+# identical encoding. The optimizer should use the histogram for estimation.
+query T retry
+SELECT info FROM [EXPLAIN SELECT * FROM t125620 WHERE ts < '2020-04-01'::TIMESTAMP] WHERE info LIKE '%estimated row count:%'
+----
+  estimated row count: 30 (30% of the table; stats collected <hidden> ago)
+
+# Verify the query still returns correct results.
+query I
+SELECT count(*) FROM t125620 WHERE ts < '2020-04-01'::TIMESTAMP
+----
+30

--- a/pkg/sql/opt_catalog.go
+++ b/pkg/sql/opt_catalog.go
@@ -2088,8 +2088,9 @@ func (os *optTableStat) init(
 			col.GetType(), string(tab.Name()), col.GetName(), stats.TSFromTime(stat.CreatedAt),
 		); err != nil {
 			// Column type in the histogram differs from column type in the
-			// table. This is only possible if we somehow re-used the same column ID
-			// during an ALTER TABLE statement, which we shouldn't.
+			// table. This can happen after a metadata-only ALTER COLUMN TYPE that
+			// changes the type family (e.g., TIMESTAMPTZ to TIMESTAMP) without
+			// rewriting data or invalidating histograms.
 			if buildutil.CrdbTestBuild {
 				return false, errors.NewAssertionErrorWithWrappedErrf(
 					err, "type check failed while initializing stat %d", stat.StatisticID,

--- a/pkg/sql/stats/histogram.go
+++ b/pkg/sql/stats/histogram.go
@@ -458,7 +458,7 @@ func (histogramData *HistogramData) TypeCheck(
 	if histogramData.ColumnType.Family() == types.BytesFamily {
 		return nil
 	}
-	if !histogramData.ColumnType.Equivalent(colType) {
+	if !hasIdenticalHistogramEncoding(histogramData.ColumnType, colType) {
 		return errors.Newf(
 			"histogram for table %v column %v created_at %s does not match column type %v: %v",
 			table, column, createdAt, colType.SQLStringForError(),
@@ -1037,4 +1037,28 @@ func expectedDistinctCount(k, n float64) float64 {
 		}
 	}
 	return count
+}
+
+// hasIdenticalHistogramEncoding returns true if values of the two types have
+// identical binary encodings for histogram bucket upper bounds. This allows
+// reusing histograms after metadata-only ALTER COLUMN TYPE conversions where
+// the underlying encoding doesn't change.
+//
+// This is specifically needed because ALTER COLUMN TYPE from TIMESTAMP to
+// TIMESTAMPTZ (or vice versa) is a metadata-only change that doesn't rewrite
+// data, but the histogram's ColumnType still references the old type. Since
+// both types use identical encoding (both store time.Time), the histogram
+// remains valid.
+func hasIdenticalHistogramEncoding(t, other *types.T) bool {
+	if t.Equivalent(other) {
+		return true
+	}
+	// TIMESTAMP and TIMESTAMPTZ have identical encoding (both store time.Time).
+	// This is the only cross-family trivial conversion defined in the schema
+	// changer where both types share the same encoding.
+	if (t.Family() == types.TimestampFamily || t.Family() == types.TimestampTZFamily) &&
+		(other.Family() == types.TimestampFamily || other.Family() == types.TimestampTZFamily) {
+		return true
+	}
+	return false
 }

